src/Common/ThreadPool.h

```diff
@@ -205,13 +207,13 @@ class ThreadFromGlobalPool : boost::noncopyable
 
     ~ThreadFromGlobalPool()
     {
-        if (joinable())
+        if (initialized())
             abort();
     }
 
     void join()
     {
-        if (!joinable())
+        if (!initialized())
```

Comment:
Maybe here we need to check for `joinable()`? I think it would be exactly the case you have described.

Reply:
Actually all places that want to join only if it is joinable should check `joinable`, and AFAIR it is true now, for example:

```
src/Access/DiskAccessStorage.cpp:    if (lists_writing_thread.joinable())
src/Access/DiskAccessStorage.cpp-        lists_writing_thread.join();
--
src/Access/DiskAccessStorage.cpp:    if (lists_writing_thread.joinable())
src/Access/DiskAccessStorage.cpp-    {
--
src/Access/ReplicatedAccessStorage.cpp:        if (watching_thread.joinable())
src/Access/ReplicatedAccessStorage.cpp-            watching_thread.join();
--
src/Client/Suggest.h:        if (loading_thread.joinable())
src/Client/Suggest.h-            loading_thread.join();
--
src/Common/Config/ConfigReloader.cpp:    if (!thread.joinable())
src/Common/Config/ConfigReloader.cpp-    {
```

The main motivation is that caller should do the check explicitly, since otherwise it is easy to "loose" the thread.
Although it will complains from dtor now.